### PR TITLE
[flutter_tools] only restrict devices based on arch + buildMode, not emulator status

### LINKED ISSUE #58886: Update emulator/simulator profile/release mode restriction to use arch restriction
Instead of attempting to gate release mode on the target device being a simulator/emulator, use a restriction based on supported AOT architectures. This would allow the tool to run release mode on x86_64 emulators today, and on iOS simulators if we decide to upload release artifacts for iOS x86_64

fyi @zanderso

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

instead of restricting profile/release mode based on whether the tool thinks the device is an emulator, restrict based on the device target architecture and the requested build mode. Notably, this enables release mode on x86_64 Android emulators, but not x86 emulators since we do not support that as an AOT target.

This does not add release mode support for simulators, since this requires us to build and upload artifacts for simulator/x86_64

Fixes https://github.com/flutter/flutter/issues/58886

Related: https://github.com/flutter/flutter/issues/11754

### CODE REVIEW

File: packages/flutter_tools/lib/src/commands/run.dart [hunk 1]
```diff
@@ -446,31 +446,15 @@ class RunCommand extends RunCommandBase {
                            'channel.', null);
     }
 
+    final BuildMode buildMode = getBuildMode();
     for (final Device device in devices) {
-      if (await device.isLocalEmulator) {
-        if (await device.supportsHardwareRendering) {
-          final bool enableSoftwareRendering = boolArg('enable-software-rendering') == true;
-          if (enableSoftwareRendering) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/run.dart:453]: TODO: re-add these checks with better implementation

File: packages/flutter_tools/lib/src/device.dart [hunk 2]
```diff
@@ -387,24 +398,17 @@ abstract class Device {
...
   Future<String> get emulatorId;
 
+  /// Whether this device can run the provided [buildMode].
+  ///
+  /// For example, some emulator architectures cannot run profile or
+  /// release builds.
+  FutureOr<bool> supportsRuntimeMode(BuildMode buildMode) => true;
+
   /// Whether the device is a simulator on a platform which supports hardware rendering.
+  @deprecated
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/device.dart:408]: TODO: remove once fixed

File: packages/flutter_tools/lib/src/device.dart [hunk 3]
```diff
@@ -581,7 +585,7 @@ abstract class Device {
         'screenshot': supportsScreenshot,
         'fastStart': supportsFastStart,
         'flutterExit': supportsFlutterExit,
-        'hardwareRendering': isLocalEmu && await supportsHardwareRendering,
+        'hardwareRendering': true,
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/device.dart:588]: TODO: remove once fixed

**Comment by jonahwilliams** [comment]
Independent of other reasons to do this - emulator detection on Android is best effort, so it lets us avoid need to keep our emulator map up to date: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/android/android_device.dart#L35 (these are physical devices that register as emulators, because samsung)

File: packages/flutter_tools/lib/src/device.dart [hunk 4]
```diff
@@ -387,24 +398,16 @@ abstract class Device {
   /// will be returned.
   Future<String> get emulatorId;
 
+  /// Whether this device can run the provided [buildMode].
+  ///
+  /// For example, some emulator architectures cannot run profile or
+  /// release builds.
+  FutureOr<bool> supportsRuntimeMode(BuildMode buildMode) => true;
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/device.dart:405]: Provided a default here for the google3 roll

**Review by zanderso — APPROVED** [review]

**Pull request merged**